### PR TITLE
mds: fix dropping events in standby replay

### PR DESCRIPTION
**Pull request by jcsp**
Ensure that we never drop the last segment during
standby replay -- this avoids the case where we
start ignoring events because we think we're
still waiting to see a subtreemap.

Fixes: http://tracker.ceph.com/issues/17954
Signed-off-by: John Spray <john.spray@redhat.com>

### CODE REVIEW

**Comment by batrick** [comment]
`Reviewed-by: Patrick Donnelly <pdonnell@redhat.com>`

File: src/mds/MDLog.cc [hunk 1]
```diff
@@ -1435,9 +1435,24 @@ void MDLog::standby_trim_segments()
   bool removed_segment = false;
   while (have_any_segments()) {
     LogSegment *seg = get_oldest_segment();
-    if (seg->end > expire_pos)
+    dout(10) << " segment seq=" << seg->seq << " " << seg->offset <<
+      "~" << seg->end - seg->offset << dendl;
+
+    if (seg->end > journaler->get_read_pos()) {
```
Comment (ukernel) [inline src/mds/MDLog.cc]: seg->end is set to journaler->get_read_pos() after reading each log event. How can this test be true?

Comment (jcsp) [inline src/mds/MDLog.cc]: You're right, I wrote this branch when experimenting and forgot to remove it.  The "segments.size() == 1" below is the part that really fixes the bug.

**Comment by jcsp** [comment]
@ukernel does it look good now?

**Comment by jcsp** [comment]
(passed tests here http://pulpito.ceph.com/jspray-2016-11-22_10:40:59-fs-wip-jcsp-testing-20161121-distro-basic-smithi/)

**Comment by ukernel** [comment]
If I understand the code correctly. The code tries trimming all segments except the head one. This seems incorrect. (If a log segment is not expired by active MDS, updates in the segment haven't
been flushed to corresponding objects. If the standby replay MDS trim the log segment, updates in the segment may get lost if the standby MDS become active later)

**Comment by jcsp** [comment]
Sorry, I messed this up when committing it.  I originally mean to remove the "seg->end > journaler->get_read_pos()" condition and leave the "seg->end > expire_pos" condition, but I removed the wrong one.  As you say, the expire_pos check still needs to be there.

**Comment by jcsp** [comment]
(it's fixed now)

**Comment by ukernel** [comment]
Reviewed-by: Yan, Zheng <zyan@redhat.com>

**Pull request merged**